test/e2e/src/util.js

```diff
@@ -91,3 +91,22 @@ async function clientFor(contract, { keypair = generateFundedKeypair(), contract
   }
 }
 module.exports.clientFor = clientFor
+
+/**
+ * Generates a ContractClient given the contractId using the from method.
+ */
+async function clientForFromTest(contractId, publicKey, keypair) {
```

Comment:
I'm in favor of replacing the guts of `clientFor` with the new `ContractClient.from` logic. Then you can skip adding any special tests for `from`, since they'll all use it. `from` already uses the constructor internally, so we're implicitly testing everything already, and then we avoid shelling out to the CLI from within `clientFor`

Reply:
Correction: we could get rid of the `xdr` lines inside the `contracts` definition. We could keep shelling out for the `contract deploy` logic, for now.

Reply:
If we do this, then we probably want to add one special test for the reverse: that a spec generated with the CLI can still work with `ContractClient`. We probably want to test it explicitly here, since the TS Bindings in the CLI will still want to use that, to avoid fetching the XDR redundantly in everyone's browser at runtime.

Reply:
Hmm, looking into this since we already have the wasmHash we should really allow the user to get the contract spec from their wasmHash since they might have that from producing it with the cli (similar to how I have it so you can get it from their wasm). So I could split the `Server.getContractWasm` into two methods: getContractWasmFromId and getContractWasmFromWasmHash or something like that. Then client.ts could use the latter if the user supplies it (getFromWasmHash). Thoughts?

Reply:
> If we do this, then we probably want to add one special test for the reverse: that a spec generated with the CLI can still work with `ContractClient`. We probably want to test it explicitly here, since the TS Bindings in the CLI will still want to use that, to avoid fetching the XDR redundantly in everyone's browser at runtime.

Makes sense

Reply:
I changed it to using fromWasmHash and added one check which uses the constructor.